### PR TITLE
Adding torchtext package

### PR DESCRIPTION
**Pull request by giswqs**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

Relevant issues:
- https://github.com/autogluon/autogluon/issues/612
- https://github.com/pytorch/text/issues/2023

### CODE REVIEW

**Comment by giswqs** [comment]
`intel-openmp` is only available for Windows.
https://anaconda.org/conda-forge/intel-openmp

**Comment by giswqs** [comment]
`torchtext` is available on the pytorch channel, but not on the conda-forge channel. 
https://anaconda.org/pytorch/torchtext

Some of the conda recipes used to build pytorch are available at https://github.com/pytorch/builder/tree/main/conda. 

I am not quite sure how to make it work for conda-forge. Any advice will be appreciated.

**Comment by dhirschfeld** [comment]
If `torchtext` is a complex compiled package you'll likely have to use a custom build script.

It looks like it's having trouble finding `cmake`
```
  running build_ext
  CMake Error: CMake was unable to find a build program corresponding to "Unix Makefiles".  CMAKE_MAKE_PROGRAM is not set.  You probably need to select a different build tool.
  CMake Error: CMAKE_C_COMPILER not set, after EnableLanguage
  CMake Error: CMAKE_CXX_COMPILER not set, after EnableLanguage
  -- Configuring incomplete, errors occurred!
```

The docs might provide useful information on how to build with `cmake`:
https://conda-forge.org/docs/maintainer/knowledge_base.html#using-cmake

You might also take a look at the `pytorch` recipe for inspiration:
https://github.com/conda-forge/pytorch-cpu-feedstock/blob/main/recipe/meta.yaml#L39-L40

**Comment by giswqs** [comment]
Making some progress. Now stuck at this [error](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=642272&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823&l=1883). I have copied the [CMakeLists.txt](https://github.com/pytorch/pytorch/blob/master/CMakeLists.txt) file from the pytorch repo, but I am not sure how to make it work here. 

>   CMake Error at third_party/CMakeLists.txt:8 (add_subdirectory):
    The source directory    /home/conda/staged-recipes/build_artifacts/torchtext_1673847419852/work/third_party/re2
    does not contain a CMakeLists.txt file.


image

**Comment by giswqs** [comment]
It appears that the sdist already contains `third_party/CMakeLists.txt`. I am not sure why the file is not accessible during build time. 

image

File: recipes/torchtext/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+outputs:
+  - name: torchtext
+    build:
+      string: cuda{{ cuda_compiler_version | replace('.', '') }}py{{ CONDA_PY }}h{{ PKG_HASH }}_{{ PKG_BUILDNUM }}  # [cuda_compiler_version != "None"]
+      string: cpu_py{{ CONDA_PY }}h{{ PKG_HASH }}_{{ PKG_BUILDNUM }}                                                # [cuda_compiler_version == "None"]
+      detect_binary_files_with_prefix: false
+      # run_exports:
+      #   - {{ pin_subpackage('pytorch', max_pin='x.x') }}
+    script: build_pytorch.sh  # [not win]
+    script: bld_pytorch.bat   # [win]
```
Comment (dhirschfeld) [inline recipes/torchtext/meta.yaml:40]: Maybe update the script names to indicate they're building `torchtext` rather than `pytorch`?

Comment (giswqs) [reply recipes/torchtext/meta.yaml]: @dhirschfeld I have changed the build script names to `torchtext` as suggested.

**Comment by giswqs** [comment]
@dhirschfeld I have added the following line to both build scripts, but still getting the same `CMakeLists.txt` file missing error. 

```bash
git submodule update --init --recursive
```

**Comment by giswqs** [comment]
I adapted the [pytorch_spline_conv](https://github.com/conda-forge/pytorch_spline_conv-feedstock/blob/main/recipe/meta.yaml) recipe suggested by @PertuyF at https://github.com/autogluon/autogluon/issues/612#issuecomment-1384282268. All the checks have passed except the `linux_64_cuda_112`,  which still throws the same `CMakeLists.txt` file missing error. Any advice? 
@malfet  @dhirschfeld @ocefpaf @BastianZim

File: recipes/torchtext/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  # https://github.com/conda-forge/conda-forge-pinning-feedstock/issues/3491
+  skip: true  # [win or cuda_compiler_version in ("None", "10.2", "11.0", "11.1")]
+  string: cpu_py{{ CONDA_PY }}h{{ PKG_HASH }}_{{ PKG_BUILDNUM }}  # [cuda_compiler_version == "None"]
+  string: cuda{{ cuda_compiler_version | replace('.', '') }}py{{ CONDA_PY }}h{{ PKG_HASH }}_{{ PKG_BUILDNUM }}  # [cuda_compiler_version != "None"]
+  script:
+    - export FORCE_CUDA="1"    # [linux64 and cuda_compiler_version != 'None']
+    - export CC="$GCC"         # [linux64 and cuda_compiler_version != 'None']
+    - export NVCC_FLAGS="--compiler-bindir=${CC}"    # [linux64 and cuda_compiler_version != 'None']
+    - git submodule update --init --recursive
+    - {{ PYTHON }} -m pip install . -vv
```
Comment (mthrok) [inline recipes/torchtext/meta.yaml:26]: torchtext has not adopted PEP-518, and I am surprised to see this works.
The official way to build is `python setup.py install`.

**Comment by giswqs** [comment]
@mthrok Thanks for the guidance at https://github.com/pytorch/text/issues/2023#issuecomment-1386446634. I will continue to explore this and see if I can make it work. 

image

**Comment by giswqs** [comment]
@ngam https://github.com/conda-forge/staged-recipes/pull/16881
@hmaarrfk https://github.com/conda-forge/staged-recipes/pull/17129

**Comment by h-vetinari** [comment]
> @h-vetinari I just realized that your PR #17129 for adding the torchtext package. I have been trying to do the same thing here. Would you be interested in collaborating on this together? I am having a hard time dealing with the torchtext third-party libraries.

I tried reviving this last December after finally adding the necessary headers to `libsentencepiece`. [However](https://github.com/conda-forge/staged-recipes/pull/17129#issuecomment-1356677971):
> (I also just found out that torchtext now depends on torchaudio, weeee)

... that package is also a [WIP](https://github.com/conda-forge/staged-recipes/pull/17082) - we made some progress with kaldi, but it still has a lot of open points. As part of the long game for kaldi-io/pykaldi, I started https://github.com/conda-forge/staged-recipes/pull/21572, which is also not trivial.

HOWEVER, I cannot find a reason in https://github.com/pytorch/text that it depends on torchaudio anymore, so I might have just managed to confuse myself. In that case, we might have a quicker path to getting torchtext sorted. :)

**Comment by h-vetinari** [comment]
> @mthrok Thanks for the guidance at [pytorch/text#2023 (comment)](https://github.com/pytorch/text/issues/2023#issuecomment-1386446634). I will continue to explore this and see if I can make it work.

You can also do this without _having_ to patch things in upstream torchtext. It could start out as local patches in conda-forge that eventually get upstreamed (carrying patches is more of a last resort, but often a reality).

**Comment by h-vetinari** [comment]
To be honest, I think we should continue in #17129, which already has revtok figured out, as well as a few other things.

Let me update the PR; I'm happy to invite you as a collaborator on my fork of staged-recipes so you can push into #17129 as well.

**Comment by giswqs** [comment]
@h-vetinari Thank you for the insights. I am closing this issue in favor of yours #17129. I have spent several days on this but still could not figure it out. Would love to collaborate with you on this.

**Pull request closed without merging**